File: leetcode/minimum-operations-to-reduce-x-to-zero.py
# 1658. Minimum Operations to Reduce X to Zero
# 🟠 Medium
#
# https://leetcode.com/problems/minimum-operations-to-reduce-x-to-zero/
#
# Tags: Array - Hash Table - Binary Search - Sliding Window - Prefix

import timeit
from typing import List


# Use two pointers to compute the maximum length subarray that has a sum
# equal to the total elements in the array minus x.
#
# Time complexity: O(n) - We visit each element of the array and do
# constant time work for each.
# Space complexity: O(1) - We use constant extra memory.
#
# Runtime 981 ms Beats 59.41%
# Memory 30.90 MB Beats 35.75%
class Solution:
    def minOperations(self, nums: List[int], x: int) -> int:
        longest, target, left, current_sum = -1, sum(nums) - x, 0, 0
        for right in range(len(nums)):
            current_sum += nums[right]
            if current_sum == target:
                longest = max(longest, 1 + right - left)
            while current_sum >= target and left < len(nums):
                current_sum -= nums[left]
                left += 1
                if current_sum == target:
                    longest = max(longest, 1 + right - left)

        return -1 if longest == -1 else len(nums) - longest


class Naive:
    def minOperations(self, nums: List[int], x: int) -> int:
        res = len(nums) + 1
        left_sum = 0
        calculations = 0
        for end in range(len(nums), 0, -1):
            if end < len(nums):
                # Add the left element value to the sum.
                left_sum += nums[end]
                if left_sum == x:
                    # If the sum of the left elements to this point
                    # matches, we cannot get a better solution later.
                    steps = len(nums) - end
                    res = min(res, steps)
                    break
                if left_sum > x:
                    # If the left sum is already greater than the
                    # number, we can't calculate a solution.
                    break

            # Now calculate sums starting from the right.
            i = 0
            # Start with whatever value we have taken from the right,
            # it will be 0 for the first iteration.
            sum = left_sum
            while i < end and sum < x:
                calculations += 1
                sum += nums[i]
                i += 1
                steps = i + (len(nums) - end)
                # If we have a match, compare the number of steps with
                # the current best.
                if sum == x:
                    res = min(res, steps)
                elif sum > x:
                    break

        return res if res < len(nums) + 1 else -1


def test():
    executors = [
        Solution,
        Naive,
    ]
    tests = [
        [[1], 1, 1],
        [[5], 5, 1],
        [[1], 3, -1],
        [[1, 1, 4, 2, 3], 5, 2],
        [[5, 6, 7, 8, 9], 4, -1],
        [[5, 6, 7, 8, 9], 35, 5],
        [[5, 6, 7, 8, 9], 30, 4],
        [[3, 2, 20, 1, 1, 3], 10, 5],
        [[3, 1, 1, 20, 1, 2, 1, 3], 27, 5],
        [[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2], 16, 15],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for col, t in enumerate(tests):
                sol = executor()
                result = sol.minOperations(t[0], t[1])
                exp = t[2]
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for"
                    + f" test {col} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/minimum-operations-to-reduce-x-to-zero.rs
// 1658. Minimum Operations to Reduce X to Zero
// 🟠 Medium
//
// https://leetcode.com/problems/minimum-operations-to-reduce-x-to-zero/
//
// Tags: Array - Hash Table - Binary Search - Sliding Window - Prefix Sum

struct Solution;
impl Solution {
    /// Use a two pointer approach to find the maximum subarray with sum equal
    /// to the sum of the values in nums - x.
    ///
    /// Time complexity: O(n) - We visit each element and do constant work for
    /// each.
    /// Space complexity: O(1) - We use constant extra memory.
    ///
    /// Runtime 18 ms Beats 40%
    /// Memory 2.85 MB Beats 80%
    pub fn min_operations(nums: Vec<i32>, x: i32) -> i32 {
        let target = nums.iter().sum::<i32>() - x;
        if target == 0 {
            return nums.len() as i32;
        }
        let (mut l, mut res, mut current_sum) = (0, 0, 0);
        for r in 0..nums.len() {
            current_sum += nums[r];
            if current_sum == target && 1 + r - l > res {
                res = 1 + r - l;
            }
            while current_sum >= target && l <= r {
                current_sum -= nums[l];
                l += 1;
                if current_sum == target && 1 + r - l > res {
                    res = 1 + r - l;
                }
            }
        }
        if res == 0 {
            -1
        } else {
            (nums.len() - res) as i32
        }
    }
}

// Tests.
fn main() {
    let tests = [
        (vec![1, 1, 4, 2, 3], 5, 2),
        (vec![5, 6, 7, 8, 9], 4, -1),
        (vec![3, 2, 20, 1, 1, 3], 10, 5),
        (
            vec![
                8828, 9581, 49, 9818, 9974, 9869, 9991, 10000, 10000, 10000, 9999, 9993, 9904,
                8819, 1231, 6309,
            ],
            134365,
            16,
        ),
    ];
    for t in tests {
        assert_eq!(Solution::min_operations(t.0, t.1), t.2);
    }
    println!("\x1b[92m» All tests passed!\x1b[0m")
}
